### PR TITLE
merge duplicate accounts which share emails

### PR DESCRIPTION
**Pull request by Xodarap**
This PR adds a migration which merges accounts that share emails using the algorithm proposed by [JP here](https://cea-core.slack.com/archives/CB5D60VSB/p1625870611062800).

Tests can be run with `yarn jest -- packages/lesswrong/server/scripts/fixDuplicateEmail.tests.ts`

### CODE REVIEW

**Comment by jpaddison3** [comment]
This is really great, thanks.

I would recommend turning off the actual mutation, and running it on the development server. Then you could see how many accounts fall into which category, and do some manual inspection. The dev server is an old clone of prod, so it should have the same data for legacy accounts.

Finally, when you run it, I'd recommend making a backup first.

File: packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    }))
+
+    // Compute a list of merges to run
+    const mergeResults = cleanUserInfo.map(user => ({
+      email: user['_id'],
+      action: mergeSingleUser(user.matches)
+    }))
+    
+    // do the merge
+    await mergeResults.filter(result => result.action != 'ManualNeeded').forEach(async result => {
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts:106]: What should happen on failures? One approach would be to have a try catch inside this loop and append the failures to a failure list. But maybe you just want to to GTFO

**Review by jpaddison3 — COMMENTED** [review]
Great work. Love The tests.

I'd like to see the results again after approval, so I'm not giving an ASTLAC, but the fixes should be pretty straightforward.

Thanks for taking this on, it's a big deal.

File: packages/lesswrong/server/scripts/fixDuplicateEmail.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
+export enum MergeType {
+  RemoveAccount,
+  KeepAccount
```
[the remark this answers is not in the document]
Comment (Xodarap) [reply packages/lesswrong/server/scripts/fixDuplicateEmail.ts:3]: Sure, what's the downside of them being numbers? Maybe I just become accustomed to this from C-like languages

File: packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      emails and includes the ids of any posts or comments they have made
+    */
+    const userInfo: Array<MongoDuplicateUser> = await Users.aggregate([
+        {
+          '$match': {
+            'deleted': false
+          }
+        }, {
+          '$group': {
+            '_id': '$email',
```
[the remark this answers is not in the document]
Comment (Xodarap) [reply packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts:28]: Good call, I transformed them to lowercase and this caught 10 additional people

File: packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xodarap) [reply packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts:106]: Oh, I did not know that about asynchronous calls. Good catch.

Yeah, given that this is a development script, I'm fine with it just dying

File: packages/lesswrong/server/scripts/fixDuplicateEmail.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  const keptUsersNames = keptUsers.map(x => x.user.matches?.username)
+  const allshareName = keptUsersNames.every(username => username == keptUsersNames[0])
+  if (allshareName) {
+    // All users have the same username, so we can merge them automatically
+
+    return {
+      destinationId: keptUsers[0].user['_id'],
+      sourceIds: classifications.map(x => x.user['_id']).filter(x => x != keptUsers[0].user['_id']),
+      justification: 'all accounts share the same name'
+    }
```
[the remark this answers is not in the document]
Comment (Xodarap) [reply packages/lesswrong/server/scripts/fixDuplicateEmail.ts:80]: Okay, I switched it to ignore spaces, underscores, and capitalization, which does automatically match a few people. But I will plan to do the rest manually.

File: packages/lesswrong/server/scripts/fixDuplicateEmail.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/scripts/fixDuplicateEmail.ts]: Harder to read in logging output. I really don't like memorizing status codes that could be strings.

File: packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+registerMigration({
+  name: "fixDuplicateEmails",
+  dateWritten: "2021-07-22",
+  idempotent: true,
+  action: async () => {
+    /*
+      This query looks complicated but it just gets all users with duplicate 
+      emails and includes the ids of any posts or comments they have made
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/migrations/2021-07-22-fixDuplicateEmails.ts]: Oh, I was just expressing excitement for the new SQL refactor. But now that you mention it, that SQL code is excellent documentation. (Which, 🤔)

**Comment by jpaddison3** [comment]
> I would recommend turning off the actual mutation, and running it on the development server. Then you could see how many accounts fall into which category, and do some manual inspection. The dev server is an old clone of prod, so it should have the same data for legacy accounts.
> 
> Finally, when you run it, I'd recommend making a backup first.

Can you also reply to this ^^^

**Comment by Xodarap** [comment]
Ah yes, I actually did my testing on the development server because that seemed easier than trying to create users myself.

And yup, making a backup first is a good thought, thanks for the wtbu!

**Review by jpaddison3 — APPROVED** [review]
LGTM

File: packages/lesswrong/server/scripts/fixDuplicateEmail.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xodarap) [reply packages/lesswrong/server/scripts/fixDuplicateEmail.ts]: Cool, that makes sense

**Comment by jpaddison3** [comment]
Arg, I missed something important. 

The `email` field is a convenience field that was a bad idea. It is supposed to reflect an email stored in `emails[n].address`, but (because mongo) is not guaranteed to. This is especially true for legacy accounts. We only log people in if they have a matching email stored in `emails[n].address`. I'm not sure (and am sleepy after a too-large lunch) what the consequences are. But note that we're potentially missing duplicate emails or merging accounts that wouldn't actually cause conflicts. (I'm guessing that the latter is more likely than the former.) Probably we should at least log `emails`.

**Pull request merged**